File: countNumberOfMaximumBitwiseORSubsets.py
from typing import List

class Solution:
    def countMaxOrSubsets(self, nums: List[int]) -> int:
        n = len(nums)
        max_or = 0
        for v in nums:
            max_or |= v

        # tail_or[i] = OR of nums[i:]
        tail_or = [0]*(n+1)
        for i in range(n-1, -1, -1):
            tail_or[i] = tail_or[i+1] | nums[i]

        def dfs(i: int, cur: int) -> int:
            if i == n:
                return 1 if cur == max_or else 0
            # prune: even with all remaining, can't reach max_or
            if (cur | tail_or[i]) < max_or:
                return 0
            return dfs(i+1, cur) + dfs(i+1, cur | nums[i])

        return dfs(0, 0)

File: countNumberOfMaximumBitwiseORSubsets.go
func countMaxOrSubsets(nums []int) int {
    n := len(nums)
    maxOr := 0
    for _, v := range nums {
        maxOr |= v
    }

    // tail[i] = OR of nums[i:]
    tail := make([]int, n+1)
    for i := n - 1; i >= 0; i-- {
        tail[i] = tail[i+1] | nums[i]
    }

    var dfs func(i, cur int) int
    dfs = func(i, cur int) int {
        if i == n {
            if cur == maxOr {
                return 1
            }
            return 0
        }
        // prune if impossible to reach maxOr
        if (cur|tail[i]) < maxOr {
            return 0
        }
        res := dfs(i+1, cur)             // skip
        res += dfs(i+1, cur|nums[i])     // take
        return res
    }

    return dfs(0, 0)
}
